pkg/util/policy/agent/agent.py

```diff
@@ -78,87 +70,73 @@ def route(self):
+        """
 
-		tid = self.json_data.get('tid')
-		logging.info("tid is %s" % tid)
+        self.http_method = "POST"
+        self.decode_request()
+        self.route()
 
-		isolated = self.json_data.get('isolated')
-		logging.info("isolated is %s" % isolated)
 
-		iptables_rules = get_current_iptables()
-		TENANT_POLICY_VECTOR_CHAIN = "ROMANA-T%s" % tid
-		logging.info("TENANT_POLICY_VECTOR_CHAIN is %s" % TENANT_POLICY_VECTOR_CHAIN)
+    def do_NP_update(self):
+        """
+        Installs or deinstalls network policy rules from iptables
+        """
+        global addr_scheme
 
-		
-		# Parse out a list of chain names out of current rules,
-		# use it to avoid duplication when adding new chains.
-		existing_chains = [ k.split(" ")[0][1:] for k in iptables_rules if k.startswith(":") ]
+        policy_def = self.json_data
+        policy_def_valid = self.validate_policy(policy_def)
+        
+        logging.warning("In do_NP_update : policy_def_valid = %s" % policy_def_valid)
 
-		tenant_vector_chain_exists = TENANT_POLICY_VECTOR_CHAIN in existing_chains
-		logging.info("tenant_vector_chain_exists is %s" % tenant_vector_chain_exists)
+        if policy_def_valid:
+            addr_scheme = policy_def.get("datacenter")
 
-		ALLOW_ANY_VECTOR = "-A %s -j ACCEPT" % TENANT_POLICY_VECTOR_CHAIN
-		logging.info("ALLOW_ANY_VECTOR is %s" % ALLOW_ANY_VECTOR)
+            self.send_response(200)
+            self.wfile.write("Policy definition accepted")
+            logging.warning("Creating policy %s" % policy_def)
 
-		allow_any_vector_exists = ALLOW_ANY_VECTOR in iptables_rules
-		logging.info("allow_any_vector_exists is %s" % allow_any_vector_exists)
+            if self.http_method == "DELETE":
+                logging.warning("Deleting policy %s" % policy_def)
+            else:
+                logging.warning("Creating policy %s" % policy_def)
 
-		filter_idx = iptables_rules.index('*filter')
+            policy_update(addr_scheme, policy_def, delete_policy = ( self.http_method == "DELETE" ))
 
-		if not tenant_vector_chain_exists:
-			logging.info("Tenant policy vector chain does not exist - creating")
-			iptables_rules.insert(filter_idx + 1, ":%s - [0:0]" % TENANT_POLICY_VECTOR_CHAIN)
+            return
+        else:
+            self.send_response(HTTP_Unprocessable_Entity)
+            self.wfile.write(""" Failed to validate policy definition """)
 
-		if allow_any_vector_exists:
-			if isolated:
-				logging.info("Enabling isolation")
-				iptables_rules.remove(ALLOW_ANY_VECTOR)
-		else:
-			if not isolated:
-				logging.info("Disabling isolation")
-				iptables_rules.insert(filter_rules_idx(iptables_rules), ALLOW_ANY_VECTOR)
-		apply_new_ruleset(iptables_rules)
 
-		return
+    def validate_policy(self, policy_def):
+        """
+        Checks that some top level sections are present in the policy.
+        """
+        policy_name = policy_def.get('name')
```

Comment:
This will give you the value for that key. The value may be empty! So, your check later on is not about whether the sections are present, but whether they are present and they also have a non-empty value.

If you just want to check that the key exists (the section exists, independent of whether there's a value or not), you can write just a single line:

    if all([ k in policy_def for k in [ "name", "applied_to", "rules", "peers", "datacenter" ]):

Otherwise, if you want to check that there are real non-empty values present for all sections, you can do this one line:

    if all([ policy_def.get(k) for k in [ "name", "applied_to", "rules", "peers", "datacenter" ]):